processing/src/main/java/org/apache/druid/query/scan/ScanQueryRunnerFactory.java

```diff
@@ -68,34 +89,224 @@ public ScanQueryRunnerFactory(
+        q.offer(srv);
+        if (q.size() > limit) {
+          q.poll();
+        }
+
+        // Finish scanning the interval containing the limit row
+        if (numRowsScanned > limit && finalInterval == null) {
+          long timestampOfLimitRow = srv.getFirstEventTimestamp(scanQuery.getResultFormat());
+          for (SegmentDescriptor descriptor : descriptorsOrdered) {
+            if (descriptor.getInterval().contains(timestampOfLimitRow)) {
+              finalInterval = descriptor.getInterval();
+            }
+          }
+          if (finalInterval == null) {
+            throw new ISE("WTH???  Row came from an unscanned interval?");
+          }
+        }
+      }
+      yielder = yielder.next(null);
+      doneScanning = yielder.isDone() ||
+                     (finalInterval != null &&
+                      !finalInterval.contains(next.getFirstEventTimestamp(scanQuery.getResultFormat())));
+    }
+    // Need to convert to a Deque because Priority Queue's iterator doesn't guarantee that the sorted order
+    // will be maintained.  Deque was chosen over list because its addFirst is O(1).
+    final Deque<ScanResultValue> sortedElements = new ArrayDeque<>(q.size());
+    while (q.size() != 0) {
+      // addFirst is used since PriorityQueue#poll() dequeues the low-priority (timestamp-wise) events first.
+      sortedElements.addFirst(q.poll());
+    }
+    return Sequences.simple(sortedElements);
+  }
+
+  @VisibleForTesting
+  Sequence<ScanResultValue> nWayMergeAndLimit(
+      List<List<QueryRunner<ScanResultValue>>> groupedRunners,
+      QueryPlus<ScanResultValue> queryPlus,
+      Map<String, Object> responseContext
+  )
+  {
+    // Starting from the innermost Sequences.map:
+    // (1) Deaggregate each ScanResultValue returned by the query runners
+    // (2) Combine the deaggregated ScanResultValues into a single sequence
+    // (3) Create a sequence of results from each runner in the group and flatmerge based on timestamp
+    // (4) Create a sequence of results from each runner group
+    // (5) Join all the results into a single sequence
+
+    return Sequences.concat(
+        Sequences.map(
+            Sequences.simple(groupedRunners),
+            runnerGroup ->
+                Sequences.map(
+                    Sequences.simple(runnerGroup),
+                    (input) -> Sequences.concat(
+                        Sequences.map(
+                            input.run(queryPlus, responseContext),
+                            srv -> Sequences.simple(srv.toSingleEventScanResultValues())
+                        )
+                    )
+                ).flatMerge(
+                    seq -> seq,
+                    Ordering.from(new ScanResultValueTimestampComparator(
+                        (ScanQuery) queryPlus.getQuery()
+                    )).reverse()
+                )
+        )
+    ).limit(
+        Math.toIntExact(((ScanQuery) (queryPlus.getQuery())).getLimit())
```

Comment:
If I do not specify a `limit` parameter but do specify an `order`, I end up in here with `limit` as `Long.MAX_VALUE` and hit a 
```
2019-03-27T23:39:05,477 ERROR [qtp904351240-69[scan_[wikiticker]_e44c2050-ecd4-45b7-ba3e-28913a46581e]] org.apache.druid.server.QueryResource - Exception handling request: {class=org.apache.druid.server.QueryResource, exceptionType=class java.lang.ArithmeticException, exceptionMessage=integer overflow, exception=java.lang.ArithmeticException: integer overflow, query=ScanQuery{dataSource='wikiticker', querySegmentSpec=MultipleSpecificSegmentSpec{descriptors=[SegmentDescriptor{interval=2015-09-12T00:00:00.000Z/2015-09-13T00:00:00.000Z, version='2019-03-08T11:26:16.569Z', partitionNumber=0}]}, virtualColumns=[], resultFormat='list', batchSize=20480, limit=9223372036854775807, dimFilter=null, columns=[], legacy=false}, peer=127.0.0.1}
java.lang.ArithmeticException: integer overflow
	at java.lang.Math.toIntExact(Math.java:1011) ~[?:1.8.0_192]
	at org.apache.druid.query.scan.ScanQueryRunnerFactory.nWayMergeAndLimit(ScanQueryRunnerFactory.java:306) ~[classes/:?]
...
```
I haven't quite figured out what the correct behavior here in that condition is yet, but this error isn't so friendly

Reply:
I'll change the `Sequence#limit()` function to accept longs.  There shouldn't be any adverse consequences.

Reply:
I _think_ that should be ok too. Should it skip the sequence limit entirely if limit is the default `Long.MAX_VALUE`?

Reply:
I don't think so because there's not really a way of distinguishing whether someone wants a limit of `Long.MAX_VALUE` or no limit at all.  Also, in practice, I don't think anyone will be querying that many rows out of Druid.

It might make more sense to do so if the default query limit value with no limit set was some invalid value like -1.